src/number_theory/legendre_symbol/add_character.lean

```diff
@@ -50,16 +52,57 @@ universes u v
 -- The domain of our additive characters
 variables (R : Type u) [add_monoid R]
 -- The target
-variables (R' : Type v) [monoid R']
+variables (R' : Type v) [comm_monoid R']
 
 /-- Define `add_char R R'` as `(multiplicative R) →* R'`.
 The definition works for an additive monoid `R` and a monoid `R'`,
 but we will restrict to the case that both are commutative rings below.
+We assume right away that `R'` is commutative, so that `add_char R R'` carries
+a structure of commutative monoid.
 The trivial additive character (sending everything to `1`) is `(1 : add_char R R').` -/
-abbreviation add_char : Type (max u v) := (multiplicative R) →* R'
+@[derive [comm_monoid, inhabited]]
+def add_char : Type (max u v) := (multiplicative R) →* R'
+
+instance add_char.has_coe_to_fun : has_coe_to_fun (add_char R R') (λ x, (multiplicative R → R')) :=
+monoid_hom.has_coe_to_fun
+
+instance add_char.monoid_hom_class : monoid_hom_class (add_char R R') (multiplicative R) R' :=
+monoid_hom.monoid_hom_class
 
 end add_char_def
 
+section group_structure
+
+universes u v
+
+namespace add_char
+
+open multiplicative
+
+variables {R : Type u} [add_comm_group R] {R' : Type v} [comm_monoid R']
+
+/-- An additive character on a commutative additive group has an inverse.
+
+Note that this is a different inverse to the one provided by `monoid_hom.has_inv`,
+as it acts on the domain instead of the codomain. -/
+instance has_inv : has_inv (add_char R R') := ⟨λ ψ, ψ.comp inv_monoid_hom⟩
+
+lemma inv_apply (ψ : add_char R R') (x : multiplicative R) : ψ⁻¹ x = ψ (of_add (- (to_add x))) :=
+rfl
+
+lemma inv_apply' (ψ : add_char R R') (x : R) : ψ⁻¹ (of_add x) = ψ (of_add (- x)) := rfl
```

Comment:
I thought I read somewhere that mathlib wants a space between operators and arguments. I'll look again...

Reply:
It's hard to tell from the sources, since statements involving `neg` are usually derived via `to_additive`... But in the docs, no space is printed, so I guess you are right.